.crawler/src/crawler.rs

```diff
@@ -0,0 +1,287 @@
+            let ping_msg = ClientMessage::Ping(
+                MESSAGE_VERSION,
+                MAXIMUM_FORK_DEPTH,
+                node.node_type(),
+                node.state(),
+                genesis.hash(),
+                Data::Object(genesis.header().clone()),
+            );
+
+            loop {
+                if node.node().num_connected() != 0 {
+                    debug!(parent: node.node().span(), "Sending out Pings");
+                    node.send_broadcast(ping_msg.clone()).unwrap();
+                }
+                tokio::time::sleep(Duration::from_secs(PING_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    /// Spawns a task dedicated to peer maintenance.
+    pub fn update_peers(&self) {
+        let node = self.clone();
+        task::spawn(async move {
+            loop {
+                // Disconnect from peers we have just crawled.
+                for addr in node.known_network.addrs_to_disconnect() {
+                    node.node().disconnect(addr).await;
+                }
+
+                // Connect to peers we haven't crawled in a while.
+                for addr in node.known_network.addrs_to_connect() {
+                    let _ = node.node().connect(addr).await;
+                }
+
+                debug!(parent: node.node().span(), "Crawling the netowrk for more peers; asking peers for their peers");
+                node.send_broadcast(ClientMessage::PeerRequest).unwrap();
+                tokio::time::sleep(Duration::from_secs(PEER_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    fn log_known_network(&self) {
+        let node = self.clone();
+        tokio::spawn(async move {
+            loop {
+                info!(parent: node.node().span(), "Known addresses: {}", node.known_network.nodes().len());
+                info!(parent: node.node().span(), "Known connections: {}", node.known_network.connections().len());
+                tokio::time::sleep(Duration::from_secs(LOG_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    /// Starts the usual periodic activities of a crawler node.
+    pub fn run_periodic_tasks(&self) {
+        self.send_pings();
+        self.update_peers();
+        self.log_known_network();
+    }
+}
+
+/// Inbound message processing logic for the crawler nodes.
+#[async_trait::async_trait]
+impl Reading for Crawler {
+    type Message = ClientMessage;
+
+    fn read_message<R: io::Read>(&self, source: SocketAddr, reader: &mut R) -> io::Result<Option<Self::Message>> {
+        // FIXME: use the maximum message size allowed by the protocol or (better) use streaming deserialization.
+        let mut buf = [0u8; 64 * 1024];
+
+        reader.read_exact(&mut buf[..MESSAGE_LENGTH_PREFIX_SIZE])?;
+        let len = u32::from_le_bytes(buf[..MESSAGE_LENGTH_PREFIX_SIZE].try_into().unwrap()) as usize;
+
+        if len > buf.len() || reader.read_exact(&mut buf[..len]).is_err() {
```

Comment:
if `len > buf.len()`, we should return `Err(io::ErrorKind::InvalidData.into())`, as this is breaking the protocol; we currently don't allocate the maximum possible size (as we will want to ignore most large messages), but anything that's above our buffer length is a no-no